Make npartitions optional when streaming from Kafka and set auto.offset.reset to latest as default.

chinmaychandak opened the pull request:
We currently need to specify the number of topic partitions (`npartitions`) to start a stream from Kafka, which is not ideal in case of production workloads where we need to add partitions to a topic on the fly, etc. The stream should be able to handle these changes on its own. No other streaming stacks need the end-user to specify #partitions for a stream. I am using Confluent Kafka's [list_topics](https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.Consumer.list_topics) API for this purpose. 

**[EDIT]** On further discussion (please read entire thread below), I have changed `npartitions` to default to `None`. I have skipped the auto-refresh capability (to support adding partitions to a topic on the fly and being handled by the stream) for now. I will add it in a follow-up PR via an additional parameter as discussed below. This PR just basically targets making the requirement of specifying number of partitions optional.

**NOTE:**
This PR also changes the default `auto.offset.reset` Kafka consumer config to `latest` conforming to the librdkafka configuration here. I needed to fix a small condition in addition to this to make sure checkpointing was working with this latest config. Currently, if a stream with `auto.offset.reset: latest` restarts, the offsets are again set to `latest` offsets (`high` watermark offset) rather than using the checkpoint created by the first run. This needed fixing.

martindurant commented:
I think this is fine, except surely there isn't a need to list the partitions on every single loop? I don't know what the overhead of that operation is, but it must add something.

> Please do not merge

Put this into the title or use a draft PR?

> I can add a test if needed

Yes please, both for increase and decrease of the partition count

chinmaychandak commented:
> I think this is fine, except surely there isn't a need to list the partitions on every single loop? I don't know what the overhead of that operation is, but it must add something.

Yeah, there would be some overhead associated, although there's two things here:

1. If we do not poll every time, and there's a partition count increase, it would cause a delay. How do we set the period at which we poll? 
2. After every poll interval, there is considerable latency reading a batch of messages — I do not think a single extra call per poll interval would add much overhead compared to that, especially since we are getting the cluster metadata for that single topic -- which would typically be smaller than a single Kafka message.

martindurant commented:
I feel like perhaps it can be optional. Agree that the user shouldn't need to specify the number, but in many cases you at least expect the number to stay static.

chinmaychandak commented:
> I feel like perhaps it can be optional.

So, basically, just set `npartitions=None` instead of `1` by default. If the user provides a value, we end up using that. If not, then use this logic?

I was also thinking of a more hacky way of implementing this. How about we set a `MAX_ALLOWED_PARTITIONS=128` and then the for loop just iterates from 1 to `MAX_ALLOWED_PARTITIONS`; the partition at which `get_watermark_offsets` throws the `KafkaException: KafkaError{code=_UNKNOWN_PARTITION,val=-190,str="Failed to get watermark offsets: Local: Unknown partition"}`, we break out of the loop and move on the next steps in the stream.

chinmaychandak commented:
> Yes please, both for increase and decrease of the partition count

I have almost never seen decreasing the partitions of a Kafka topic. It's really hard for Kafka to migrate data and not recommended. You generally only increase. If you need to decrease, it's better to delete the topic and create a new one.

chinmaychandak commented:
What do you think of the hacky solution @martindurant?



> So, basically, just set `npartitions=None` instead of `1` by default. If the user provides a value, we end up using that. If not, then use this logic? I still feel removing `npartitions` would be better in the long run, though.

Do you feel this is the correct way of making it optional? Although I feel it would be overcomplicating things.

martindurant commented:
Yes, I think that allowing specifying the number of partitions is OK, and doing a list the first time to get that number if passed None. *In addition*, there can be an option to check the number every N cycles (or every Y seconds).

chinmaychandak commented:
@martindurant Makes sense. 

I have changed `npartitions` to default to `None`. Also, I have skipped the auto-refresh capability of the number of partitions for now. I will add it in a follow-up PR via an additional parameter.

I have also changed the default `auto.offset.reset` Kafka consumer config to `latest` conforming to the librdkafka configuration [here](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md). 

I needed to fix a small condition in addition to this to make sure checkpointing was working with this `latest` config. Currently, if a stream with `auto.offset.reset: latest` restarts, the offsets are again set to `latest` offsets (high watermark offset) rather than using the checkpoint created by the first run. This needed fixing.

Let me know if this looks good. We are still working on adding the `list_topics()` API to custreamz, hopefully the DRAFT tag can be removed from this PR early next week.

martindurant commented:
Yes, I think this is along the right lines.

> I have also changed the default auto.offset.reset Kafka consumer config to latest conforming to the librdkafka configuration here.

This should be documented for the source class. I think most other sources (like file) will start from the earliest data.

chinmaychandak commented:
> This should be documented for the source class. 

Done

chinmaychandak commented:
@martindurant This is ready for review now, I have removed the DRAFT label. Apologies for the delay, unexpected hiccups slowed this down.

CJ-Wright commented on setup.py:
```diff
@@ -9,7 +9,7 @@
 
 setup(name='streamz',
-      version='0.5.6',
+      version='0.5.5',
```
?

martindurant commented on streamz/sources.py:
```diff
@@ -510,7 +517,9 @@ def checkpoint_emit(_part):
                     if 'auto.offset.reset' in self.consumer_params.keys():
-                        if self.consumer_params['auto.offset.reset'] == 'latest':
+                        if self.consumer_params['auto.offset.reset'] == 'latest' and \
+                                (self.positions == [-1001] * self.npartitions
```
Where does -1001 come from?
In this expression, I would prefer parentheses to make clear which `and`/`or` goes together.

chinmaychandak replied:
> Where does -1001 come from?

The default value returned by `consumer.committed()` which is assigned to `self.positions` is `-1001`. 

>  parentheses to make clear which `and`/`or` goes together

There's one already there. It shows `latest` `and` (`-1001` `or` `0`). Did you mean another parenthesis around the entire expression?

martindurant commented on streamz/tests/test_kafka.py:
```diff
@@ -223,6 +223,7 @@ def test_kafka_batch():
         stream.upstream.stopped = True
 
 
+@flaky(max_runs=3, min_passes=1)
```
Any idea why? We would rather not have this. How frequently does it fail?

chinmaychandak replied:
Nope, no idea at all. This was really surprising for me, too. Unlike other flaky tests, this one passes >90% of the time, but I did see it fail one time, and succeed the next. Hence, I thought adding `flaky` would be best. Should I remove it?

martindurant replied:
Do you believe it was already failing sometimes before, or did something change?

chinmaychandak replied:
Maybe it was a one-off, although it is concerning.

chinmaychandak replied:
I don't think anything changed, but I did recall Jarod saying that it failed for him once too, and this was way before the checkpointing stuff.

martindurant replied:
I would be tempted to leave it as was, and see how often this arises. We can investigate at a later date whats going on - I don't suppose we even know at what stage things are hanging.

chinmaychandak replied:
Gotcha, will remove the flaky marker then.

chinmaychandak replied:
Actually, I'm running all tests locally, and this test seems to fail like 1 out of 5 times.

martindurant commented on streamz/tests/test_kafka.py:
```diff
@@ -291,11 +294,13 @@ def test_kafka_dask_checkpointing_sync_nodes(c, s, w1, w2):
             'group.id': 'streamz-test%i' % j2,
-            'enable.auto.commit': False}
+            'enable.auto.commit': False,
+            'auto.offset.reset': 'earliest'}
```
We could do with some more tests, e.g., with different number of partitions and whether the user passes npartitions or not (and what if the number passed is larger or smaller than that set in kafka); and the offsets earliest/latest. 

This checkpointing test is I suppose a separate thing, but it would be nice to see all the combinations tested somewhere.

chinmaychandak replied:
> with different number of partitions

Yes, we need this test.

> whether the user passes npartitions or not

We already have this, correct? But I guess none of the tests have the `npartitions` parameter specified  (i.e., all of them use the new default value of `None`), which I think is ok. But yeah, we can add one with `npartitions` maybe specified to something >1. 

> and what if the number passed is larger or smaller than that set in kafka

Not sure I understand. The support for adding dynamic partitions on the fly would be a follow-up PR like we discussed (auto-refresh partitions every `N` cycles based on a user parameter).

 > offsets earliest/latest

These have been covered extensively, although they are spread across different tests. 

Should I add a test with `npartitions` specified and >1 for now? Or can it be in a separate PR with the adding partitions on the fly capability?

martindurant replied:
I think certainly we should add npartitions=-1, 0, 1, 2, 4 as a set of tests where the number of real partitions in kafka is 2 (fixed). The first two should fail with ValueError, the middle one should get about half the data, and the last two should both get all the data, but the last one will have partitions that never receive anything. Does that sound right?

chinmaychandak replied:
For npartitions=-1/0, it doesn't enter the polling consumption [loop](https://github.com/chinmaychandak/streamz/blob/ccc7d428ab24939fe07b151a7469a335f853ef9b/streamz/sources.py#L512). 

For npartitions=1, yes, about half the data. 

For npartitions=2/4, yes, they get all the data, but in the last one, partitions 3 & 4 are skipped in every polling consumption loop because the `get_watermarks_offset` throws an [error](https://github.com/chinmaychandak/streamz/blob/ccc7d428ab24939fe07b151a7469a335f853ef9b/streamz/sources.py#L517). 

All of these are handled within `from_kafka_batched` gracefully IMO (we can add a check at the beginning itself that raises a ValueError if npartitions is <=0, if need be), so we might not need tests. However, I do think we need a test with `npartitions` specified explicitly with a value `>1`, and maybe it can be plugged inside the test with the auto-refresh capability.

What do you think?

chinmaychandak replied:
Added some tests, do have a look

chinmaychandak replied:
@martindurant, please do let me know if we need to add anything else to this to get it rolled in. If not, there's another PR for auto-refreshing partitions on the way! :)

martindurant commented on streamz/tests/test_kafka.py:
```diff
@@ -61,6 +50,7 @@ def stop_docker(name='streamz-kafka', cid=None, let_fail=False):
 
 def launch_kafka():
     stop_docker(let_fail=True)
+    os.system("docker pull spotify/kafka")
```
Strictly, docker should do this when you ask to run on the next line, and `os.system` is [discouraged](https://docs.python.org/3.8/library/subprocess.html#replacing-os-system) versus `subprocess.call` here and below.
It would look like (NB: better execute command directly than spawn a shell)
```
subprocess.call(shlex.split("docker pull spotify/kafka"))
```
(if indeed you want to keep the extra line, which is fine)

chinmaychandak replied:
Good point, I didn't know about this. Thanks for pointing this out!

martindurant commented on streamz/tests/test_kafka.py:
```diff
@@ -244,6 +234,63 @@ def test_kafka_dask_batch(c, s, w1, w2):
+                kafka.produce(TOPIC, b'value-%d' % i, partition=1)
+        kafka.flush()
+
+        try:
```
```python
with pytest.raises(ValueError):
```

chinmaychandak replied:
Doh, don't know how I missed this!

martindurant reviewed (approved):
Sorry, just a couple of style requests, then good to go.

martindurant commented:
Merging now - travis is not reporting the test success for some reason.
